scalameta/dialects/shared/src/main/scala/scala/meta/Dialect.scala

```diff
@@ -0,0 +1,427 @@
+    privateCopy(allowEnums = allowEnums)
+  }
+  def withAllowImplicitByNameParameters(allowImplicitByNameParameters: Boolean): Dialect = {
+    privateCopy(allowImplicitByNameParameters = allowImplicitByNameParameters)
+  }
+  def withAllowImplicitFunctionTypes(allowImplicitFunctionTypes: Boolean): Dialect = {
+    privateCopy(allowImplicitFunctionTypes = allowImplicitFunctionTypes)
+  }
+  def withAllowInlineIdents(allowInlineIdents: Boolean): Dialect = {
+    privateCopy(allowInlineIdents = allowInlineIdents)
+  }
+  def withAllowInlineMods(allowInlineMods: Boolean): Dialect = {
+    privateCopy(allowInlineMods = allowInlineMods)
+  }
+  def withAllowLiteralTypes(allowLiteralTypes: Boolean): Dialect = {
+    privateCopy(allowLiteralTypes = allowLiteralTypes)
+  }
+  def withAllowMethodTypes(allowMethodTypes: Boolean): Dialect = {
+    privateCopy(allowMethodTypes = allowMethodTypes)
+  }
+  def withAllowMultilinePrograms(allowMultilinePrograms: Boolean): Dialect = {
+    privateCopy(allowMultilinePrograms = allowMultilinePrograms)
+  }
+  def withAllowOrTypes(allowOrTypes: Boolean): Dialect = {
+    privateCopy(allowOrTypes = allowOrTypes)
+  }
+  def withAllowPatUnquotes(allowPatUnquotes: Boolean): Dialect = {
+    privateCopy(allowPatUnquotes = allowPatUnquotes)
+  }
+  def withAllowSpliceUnderscores(allowSpliceUnderscores: Boolean): Dialect = {
+    privateCopy(allowSpliceUnderscores = allowSpliceUnderscores)
+  }
+  def withAllowTermUnquotes(allowTermUnquotes: Boolean): Dialect = {
+    privateCopy(allowTermUnquotes = allowTermUnquotes)
+  }
+  def withAllowToplevelTerms(allowToplevelTerms: Boolean): Dialect = {
+    privateCopy(allowToplevelTerms = allowToplevelTerms)
+  }
+  def withAllowTrailingCommas(allowTrailingCommas: Boolean): Dialect = {
+    privateCopy(allowTrailingCommas = allowTrailingCommas)
+  }
+  def withAllowTraitParameters(allowTraitParameters: Boolean): Dialect = {
+    privateCopy(allowTraitParameters = allowTraitParameters)
+  }
+  def withAllowTypeLambdas(allowTypeLambdas: Boolean): Dialect = {
+    privateCopy(allowTypeLambdas = allowTypeLambdas)
+  }
+  def withAllowViewBounds(allowViewBounds: Boolean): Dialect = {
+    privateCopy(allowViewBounds = allowViewBounds)
+  }
+  def withAllowWithTypes(allowWithTypes: Boolean): Dialect = {
+    privateCopy(allowWithTypes = allowWithTypes)
+  }
+  def withAllowXmlLiterals(allowXmlLiterals: Boolean): Dialect = {
+    privateCopy(allowXmlLiterals = allowXmlLiterals)
+  }
+  def withToplevelSeparator(toplevelSeparator: String): Dialect = {
+    privateCopy(toplevelSeparator = toplevelSeparator)
+  }
+  def withAllowNumericLiteralUnderscoreSeparators(
+      allowNumericLiteralUnderscoreSeparators: Boolean
+  ): Dialect = {
+    privateCopy(allowNumericLiteralUnderscoreSeparators = allowNumericLiteralUnderscoreSeparators)
+  }
+
+  // NOTE(olafur): add the next `withX()` method above this comment.
+
+  private[this] def privateCopy(
```

Comment:
interesting idea... why not make this public, and get rid of all those `withXxx` methods? this would look exactly like a `case class` copy, and if anyone ever used `.copy` elsewhere, would continue to work unchanged.

Reply:
If you used `copy` method that defines all parameters, on every new parameter added you would break binary compatibility. If you only use `withX` those methods don't change with addition of new parameters, you are safe. (My reasoning for this change :P also binary compatibility sucks :< )

Reply:
a, rozumiem...

Reply:
@kpbochenek Correct, by keeping this method private we don't break the public API when we add new parameters.

Reply:
@kpbochenek I agree binary compatibility can be annoying, but it could be worse. My understanding is that evolving APIs in a binary compatible way is a lot harder and sometimes near impossible in other languages/ecosystems. It's at least **possible** with Scala/JVM, even if it's cumbersome.